File: app/src/main/java/dk/huskplakaten/app/AuthSessionStore.kt
package dk.huskplakaten.app

import android.content.Context
import java.util.UUID

data class AuthSession(
    val userId: String,
    val email: String,
    val billingActive: Boolean
)

class AuthSessionStore(context: Context) {
    private val prefs = context.getSharedPreferences("auth_session", Context.MODE_PRIVATE)

    fun restore(): AuthSession? {
        val userId = prefs.getString(KEY_USER_ID, null) ?: return null
        val email = prefs.getString(KEY_EMAIL, null) ?: return null
        val billingActive = prefs.getBoolean(KEY_BILLING_ACTIVE, false)
        return AuthSession(userId = userId, email = email, billingActive = billingActive)
    }

    fun loginWithEmail(email: String, password: String): Result<AuthSession> {
        if (!email.contains("@")) return Result.failure(IllegalArgumentException("Ugyldig email."))
        if (password.length < 6) return Result.failure(IllegalArgumentException("Kodeord skal mindst være 6 tegn."))
        val userId = "email_${UUID.nameUUIDFromBytes(email.lowercase().toByteArray())}"
        return Result.success(
            AuthSession(
                userId = userId,
                email = email.trim(),
                billingActive = prefs.getBoolean(KEY_BILLING_ACTIVE, false)
            ).also { save(it) }
        )
    }

    fun loginWithGooglePlaceholder(): AuthSession {
        val email = "google.user@placeholder.local"
        val session = AuthSession(
            userId = "google_${UUID.randomUUID()}",
            email = email,
            billingActive = prefs.getBoolean(KEY_BILLING_ACTIVE, false)
        )
        save(session)
        return session
    }

    fun setBillingActive(active: Boolean): AuthSession? {
        val existing = restore() ?: return null
        val updated = existing.copy(billingActive = active)
        save(updated)
        return updated
    }

    fun clear() {
        prefs.edit().clear().apply()
    }

    private fun save(session: AuthSession) {
        prefs.edit()
            .putString(KEY_USER_ID, session.userId)
            .putString(KEY_EMAIL, session.email)
            .putBoolean(KEY_BILLING_ACTIVE, session.billingActive)
            .apply()
    }

    companion object {
        private const val KEY_USER_ID = "user_id"
        private const val KEY_EMAIL = "email"
        private const val KEY_BILLING_ACTIVE = "billing_active"
    }
}

File: lib/auth/auth_session_store.dart
import 'dart:convert';

import 'package:shared_preferences/shared_preferences.dart';

class AuthSession {
  const AuthSession({
    required this.userId,
    required this.email,
    required this.billingActive,
  });

  final String userId;
  final String email;
  final bool billingActive;

  AuthSession copyWith({
    String? userId,
    String? email,
    bool? billingActive,
  }) {
    return AuthSession(
      userId: userId ?? this.userId,
      email: email ?? this.email,
      billingActive: billingActive ?? this.billingActive,
    );
  }

  Map<String, dynamic> toMap() {
    return <String, dynamic>{
      'userId': userId,
      'email': email,
      'billingActive': billingActive,
    };
  }

  factory AuthSession.fromMap(Map<String, dynamic> map) {
    return AuthSession(
      userId: map['userId'] as String? ?? '',
      email: map['email'] as String? ?? '',
      billingActive: map['billingActive'] as bool? ?? false,
    );
  }
}

class AuthSessionStore {
  static const String _sessionKey = 'auth_session_v1';

  Future<AuthSession?> restore() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(_sessionKey);
    if (raw == null || raw.isEmpty) {
      return null;
    }
    try {
      final decoded = jsonDecode(raw) as Map<String, dynamic>;
      final session = AuthSession.fromMap(decoded);
      if (session.userId.isEmpty || session.email.isEmpty) {
        return null;
      }
      return session;
    } catch (_) {
      return null;
    }
  }

  Future<AuthSession> loginWithEmail(String email, String password) async {
    if (!email.contains('@')) {
      throw Exception('Ugyldig email.');
    }
    if (password.length < 6) {
      throw Exception('Kodeord skal mindst være 6 tegn.');
    }

    final normalizedEmail = email.trim();
    final userId = 'email_${normalizedEmail.toLowerCase()}';
    final existing = await restore();
    final session = AuthSession(
      userId: userId,
      email: normalizedEmail,
      billingActive: existing?.billingActive ?? false,
    );
    await _save(session);
    return session;
  }

  Future<AuthSession> loginWithGooglePlaceholder() async {
    final existing = await restore();
    final session = AuthSession(
      userId: 'google_${DateTime.now().millisecondsSinceEpoch}',
      email: 'google.user@placeholder.local',
      billingActive: existing?.billingActive ?? false,
    );
    await _save(session);
    return session;
  }

  Future<AuthSession?> setBillingActive(bool active) async {
    final existing = await restore();
    if (existing == null) {
      return null;
    }
    final updated = existing.copyWith(billingActive: active);
    await _save(updated);
    return updated;
  }

  Future<void> clear() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_sessionKey);
  }

  Future<void> _save(AuthSession session) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_sessionKey, jsonEncode(session.toMap()));
  }
}